File: iosApp/iosApp/Features/Transactions/TransactionListViewModel.swift
import Foundation
import Shared

/// 取引履歴画面の表示状態。
enum TransactionListUiState {
    case loading
    case loaded(transactions: [Shared.Transaction], refreshing: Bool)
    case error(message: String)
}

/// 取引履歴画面の状態とアクションをまとめる ViewModel。
///
/// shared 側 `TransactionsFlowIos.fetchMyTransactions` を呼んで取引一覧を取得する。
@MainActor
final class TransactionListViewModel: ObservableObject {
    @Published private(set) var state: TransactionListUiState = .loading

    private let userRepository: UserRepository
    private let sessionStore: SessionStore
    private var inFlight: Kotlinx_coroutines_coreJob?

    init() {
        self.userRepository = KoinIosKt.userRepository()
        self.sessionStore = KoinIosKt.sessionStore()
    }

    deinit {
        // deinit は @MainActor 隔離外で実行されるが、Kotlinx.coroutines の Job.cancel は
        // thread-safe なため、ここから直接 cancel しても安全。
        inFlight?.cancel(cause: nil)
    }

    func onAppear() {
        if case .loading = state {
            load(initial: true)
        }
    }

    func refresh() {
        load(initial: false, completion: nil)
    }

    /// `.refreshable` 用の async 版 refresh。fetch 完了まで待つことで、引きおろしジェスチャの
    /// インジケータがジョブ完了前に閉じないようにする。
    func refreshAwait() async {
        await withCheckedContinuation { (continuation: CheckedContinuation<Void, Never>) in
            load(initial: false) {
                continuation.resume()
            }
        }
    }

    private func load(initial: Bool, completion: (@MainActor () -> Void)? = nil) {
        inFlight?.cancel(cause: nil)
        if !initial, case let .loaded(items, _) = state {
            state = .loaded(transactions: items, refreshing: true)
        }
        inFlight = TransactionsFlowIosKt.fetchMyTransactions(
            userRepository: userRepository,
            sessionStore: sessionStore,
        ) { [weak self] outcome in
            Task { @MainActor in
                guard let self else {
                    completion?()
                    return
                }
                switch outcome {
                case let loaded as TransactionsLoadOutcome.Loaded:
                    self.state = .loaded(transactions: loaded.transactions, refreshing: false)
                case let failure as TransactionsLoadOutcome.Failure:
                    self.state = .error(message: failure.message)
                case let netFailure as TransactionsLoadOutcome.NetworkFailure:
                    self.state = .error(message: netFailure.message)
                case is TransactionsLoadOutcome.Unauthenticated:
                    self.state = .error(message: "セッションが切れました")
                default:
                    self.state = .error(message: "未知のエラーが発生しました")
                }
                completion?()
            }
        }
    }
}

File: composeApp/src/androidMain/kotlin/studio/nxtech/fujubank/features/transactions/TransactionListViewModel.kt
package studio.nxtech.fujubank.features.transactions

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import studio.nxtech.fujubank.data.remote.NetworkResult
import studio.nxtech.fujubank.data.repository.UserRepository
import studio.nxtech.fujubank.session.SessionState
import studio.nxtech.fujubank.session.SessionStore

/**
 * 取引履歴画面の状態とアクションを束ねる ViewModel。
 *
 * - 起動時に [SessionStore] から自分の `userId` を取得して取引一覧を fetch する。
 * - `refresh()` で再取得（pull-to-refresh 用）。
 *
 * MVP では pagination を持たず、サーバー側のデフォルト件数を全件表示する。
 */
class TransactionListViewModel(
    private val userRepository: UserRepository,
    private val sessionStore: SessionStore,
) : ViewModel() {

    private val _state = MutableStateFlow<TransactionListUiState>(TransactionListUiState.Loading)
    val state: StateFlow<TransactionListUiState> = _state.asStateFlow()

    // 進行中の fetch Job。新しい load() で前回をキャンセルして後勝ち上書きを防ぐ。
    private var loadJob: Job? = null

    init {
        load(initial = true)
    }

    fun refresh() {
        load(initial = false)
    }

    private fun load(initial: Boolean) {
        loadJob?.cancel()
        if (!initial) {
            _state.update { current ->
                if (current is TransactionListUiState.Loaded) current.copy(refreshing = true) else current
            }
        }
        loadJob = viewModelScope.launch {
            val sessionUserId = (sessionStore.current as? SessionState.Authenticated)?.userId
            // ダミーモードでは Repository が userId を無視してフェイクデータを返すため、
            // セッション未確立でも UI 確認のために空文字でフォールスルーさせる。
            if (sessionUserId == null && !userRepository.useDummyData) {
                _state.value = TransactionListUiState.Error(message = "セッションが切れました")
                return@launch
            }
            val userId = sessionUserId ?: ""
            when (val result = userRepository.transactions(userId)) {
                is NetworkResult.Success -> _state.value = TransactionListUiState.Loaded(
                    items = result.value.sortedByDescending { it.occurredAt },
                )
                is NetworkResult.Failure -> _state.value = TransactionListUiState.Error(
                    message = "取引履歴を取得できませんでした",
                )
                is NetworkResult.NetworkFailure -> _state.value = TransactionListUiState.Error(
                    message = "通信エラーが発生しました",
                )
            }
        }
    }
}
